File: Data Structures & Algorithms/products-of-array-discluding-self/submission-3.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        prefix_arr = [1]
        sufix_arr = [1]
        for i in range(len(nums) - 1):
            prefix_arr.append(prefix_arr[-1] * nums[i])
            sufix_arr.append(sufix_arr[-1] * nums[-i - 1])
        
        sufix_arr = sufix_arr[::-1]

        result = [sufix_arr[j] * prefix_arr[j] for j in range(len(nums))]
        return result

File: Data Structures & Algorithms/subsets/submission-3.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        List<Integer> path = new ArrayList<>();

        backtrack(nums, 0, path, result);
        return result;
    }
    private void backtrack(int[] nums, int start, List<Integer>path, List<List<Integer>>result) {
        result.add(new ArrayList<>(path));
        
        for (int i = start; i < nums.length; i ++) {
            path.add(nums[i]);
            backtrack(nums, i + 1, path, result);
            path.remove(path.size() - 1);
        }
    }
}
